packages/cog/src/cog/job.cutline.ts

```diff
@@ -1,85 +1,76 @@
      */
-    constructor(geojson?: FeatureCollection, maxZoom = 13) {
-        this.maxZoom = maxZoom;
+    constructor(geojson?: FeatureCollection) {
         if (geojson == null) return;
         if (findGeoJsonProjection(geojson) !== EPSG.Wgs84) {
             throw new Error('Invalid geojson; CRS may not be set for cutline!');
         }
         for (const { geometry } of geojson.features) {
             if (geometry.type === 'MultiPolygon') {
                 for (const [coords] of geometry.coordinates) {
-                    this.addPolygon(coords);
+                    this.polygons.add(coords);
                 }
             } else if (geometry.type === 'Polygon') {
-                this.addPolygon(geometry.coordinates[0]);
+                this.polygons.add(geometry.coordinates[0]);
             }
         }
     }
 
-    addPolygon(coords: Position[]): void {
-        const polygon = GeoJson.toFeaturePolygon([coords]) as Feature<Polygon>;
-        this.polygons.add(coords);
-        const covering = TileCover.cover(polygon.geometry, 1, this.maxZoom);
-        for (const qkey of covering) {
-            this.trie.add(qkey);
-            addPolygonToQuadKey(this, qkey, coords);
+    /**
+     * Find the covering of source imagery and a (optional) cutline. Truncates the cutline to match.
+
+     * @param sourceBounds geoJson of source imagery
+     * @param coverZ zoom level to cover at
+     * @return the covering at sourceZ resolution
+     */
+    filter(sourceBounds: GeoJSON.FeatureCollection, coverZ: number): QuadKeyTrie {
+        const covering = new QuadKeyTrie();
+
+        for (const { geometry } of sourceBounds.features) {
+            if (geometry.type === 'Polygon') {
+                mergeCovering(covering, TileCover.cover(geometry, 1, coverZ));
+            }
         }
-        let prevQuadKey;
-        for (let i = 0; i < coords.length; ++i) {
-            const point = coords[i];
-            const quadKeys = this.trie.getPoint(point);
-            for (const q of quadKeys) {
-                if (q != prevQuadKey) {
-                    prevQuadKey = q;
-                    const cm = this.quadKeyMap.get(q)!;
-                    if (cm.has(coords)) {
-                        cm.set(coords, i);
-                        for (const bounds = Bounds.fromQuadKey(q); i + 1 < coords.length; ++i) {
-                            if (!bounds.containsPoint(coords[i + 1])) break;
-                        }
-                        break;
-                    }
-                }
+
+        if (this.polygons.size == 0) return covering;
+
+        // We have a cutline. Filter it by sourceBounds and return intersection
+        const keepLines = new Set<GeoJSON.Position[]>();
+
+        const coords: Position[][] = [[]];
+
+        const geom = GeoJson.toFeaturePolygon(coords).geometry as GeoJSON.Polygon;
+
+        const clCovering = new QuadKeyTrie();
+
+        for (const poly of this.polygons) {
+            coords[0] = poly;
```

Comment:
Just make a new geojson, screwing with the internal coordinate array makes this code block difficult to understand.

readability is much better than performance gain you would get by doing this.